sklearn/preprocessing/discretization.py

```diff
@@ -0,0 +1,228 @@
+        *Y*: A python list of numeric class labels.
+
+        Returns a list of intervals, based upon the MDLP stopping criterion.
+        """
+
+        order = np.argsort(attributes)
+        x = attributes[order]
+        y = Y[order]
+
+        intervals = list()
+
+        def get_cut(ind):
+            return (x[ind-1] + x[ind]) / 2
+
+        # Now we do a depth first search to create intervals
+        num_samples = len(x)
+        search_intervals = list()
+        search_intervals.append((0, num_samples, 0))
+        while len(search_intervals) > 0:
+            start, end, depth = search_intervals.pop()
+
+            k = self._find_cut(y, start, end)
+
+            # Need to see whether the "front" and "back" of the intervals need
+            # to be float("-inf") or float("inf")
+            if (k == -1) or (self._reject_split(y, start, end, k) and
+                             depth >= self.min_depth):
+                front = float("-inf") if (start == 0) else get_cut(start)
+                back = float("inf") if (end == num_samples) else get_cut(end)
+
+                # Corner case: If front == back, don't add it in
+                if front < back:
+                    intervals.append((front, back))
+                continue
+
+            search_intervals.append((start, k, depth + 1))
+            search_intervals.append((k, end, depth + 1))
+
+        return intervals
+
+    @staticmethod
+    def _slice_entropy(y, start, end):
+        """Returns the entropy of the given slice of y. Also returns the
+        number of classes within the interval.
+        """
+        counts = np.bincount(y[start:end])
+        vals = counts / (end - start)
+        return entropy(vals), np.sum(vals != 0)
+
+    @staticmethod
+    def _reject_split(y, start, end, k):
+        """Using the minimum description length principal, determines
+        whether it is appropriate to stop cutting.
+        """
+
+        N = end - start
+        entropy1, k1 = MDLP._slice_entropy(y, start, k)
+        entropy2, k2 = MDLP._slice_entropy(y, k, end)
+        whole_entropy, k = MDLP._slice_entropy(y, start, end)
+
+        # Calculate the final values
+        gain = whole_entropy - 1 / N * ((start - k) * entropy1 +
+                                        (end - k) * entropy2)
+        entropy_diff = k * whole_entropy - k1 * entropy1 - k2 * entropy2
+        delta = log(3**k - 2) - entropy_diff
+
+        return gain <= 1 / N * (log(N - 1) + delta)
+
+    @staticmethod
+    def _find_cut(y, start, end):
+        """Finds the best cut between the specified interval.
+
+        Returns a candidate cut point index k. k == -1 if there is no good cut,
+        otherwise k is in {1, ..., len(x) - 2}
+        """
+
+        # Want to probe for the best partition _entropy in a "smart" way
+        # Input is the splitting index, to create partitions [start, ind)
+        # and [ind, end).
+        length = end - start
+
+        def partition_entropy(ind):
```

Comment:
Please inline this function. It doesn't help to be given a name, given that it is called once.